tests/mpi/hp_step-4.cc

```diff
@@ -0,0 +1,352 @@
+
+    hp::FEValues<dim> hp_fe_v(dof_handler.get_fe_collection(),
+                              q_collection,
+                              update_JxW_values | update_gradients |
+                                update_values | update_quadrature_points);
+
+    std::vector<types::global_dof_index> local_dof_indices;
+
+    FullMatrix<double> cell_matrix;
+    Vector<double>     cell_rhs;
+
+    for (auto &cell : dof_handler.active_cell_iterators())
+      {
+        if (cell->is_locally_owned() && cell->active_fe_index() == 0)
+          {
+            const unsigned int active_fe_index = cell->active_fe_index();
+
+            if (active_fe_index == 0)
+              {
+                hp_fe_v.reinit(cell);
+
+                const FEValues<dim> &fe_values =
+                  hp_fe_v.get_present_fe_values();
+
+                cell_matrix.reinit(fe[active_fe_index].dofs_per_cell,
+                                   fe[active_fe_index].dofs_per_cell);
+
+                cell_rhs.reinit(fe[active_fe_index].dofs_per_cell);
+
+                local_dof_indices.resize(fe[active_fe_index].dofs_per_cell);
+
+                cell->get_dof_indices(local_dof_indices);
+
+                const unsigned int dofs_per_cell =
+                  fe[active_fe_index].dofs_per_cell;
+
+                const unsigned int n_q_points =
+                  q_collection[active_fe_index].size();
+
+                for (unsigned int q_index = 0; q_index < n_q_points; ++q_index)
+                  for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                    {
+                      for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                        cell_matrix(i, j) += (fe_values.shape_grad(i, q_index) *
+                                              fe_values.shape_grad(j, q_index) *
+                                              fe_values.JxW(q_index));
+
+                      const auto x_q = fe_values.quadrature_point(q_index);
+                      cell_rhs(i) +=
+                        (fe_values.shape_value(i, q_index) *
+                         right_hand_side.value(x_q) * fe_values.JxW(q_index));
+                    }
+
+                constraints.distribute_local_to_global(
+                  cell_matrix, cell_rhs, local_dof_indices, system_matrix, rhs);
+              }
+          }
+      }
+    system_matrix.compress(VectorOperation::add);
+
+    rhs.compress(VectorOperation::add);
+  }
+
+
+
+  template <int dim>
+  void
+  Step4<dim>::solve()
+  {
+    SolverControl solver_control(system_matrix.m(), 1e-12 * rhs.l2_norm());
+
+    TrilinosWrappers::SolverDirect::AdditionalData add_data(false,
+                                                            "Amesos_Mumps");
+
+    TrilinosWrappers::SolverDirect direct_solver(solver_control, add_data);
+
+    TrilinosWrappers::MPI::Vector completely_distributed_solution(
+      locally_owned_dofs, communicator);
+
+    direct_solver.solve(system_matrix, completely_distributed_solution, rhs);
```

Comment:
It would probably be a good idea to also check for the number of steps the solver takes. Depending on the version of Trilinos, this number may vary. We should thus check for a range of values.

There is a macro defined in `tests.h` that does exactly this and can be used like the following:
```
check_solver_within_range(direct_solver.solve(system_matrix,
                                              completely_distributed_solution,
                                              rhs),
                          solver_control.last_step(),
                          /*min_allowed=*/10,
                          /*max_allowed=*/80);
```

Reply:
Is that necessary for the direct solver?

Reply:
You are correct, this only makes sense for iterative solvers. I got so used to iterative solvers and the `check_solver_within_range` function that I didn't even think of this :)